### DISCUSSION TITLE
Parsing and generating JSON values

### DISCUSSION THREAD

**Discussion by doolio** [joaotavora/eglot#1084] [Q&A] [answered]
The [manual](https://github.com/joaotavora/eglot/blob/master/MANUAL.md#format-of-the-value) states the following:

> In any case, the JSON values `true`, `false` and `{}` are represented by the Elisp values `t`, `:json-false` and `nil`, respectively.

But what about the JSON `null` and `[]` values?  The manual makes no mention of these. Are they not used by any LSP server?

So according to the manual the symbol `nil` represents the empty JSON object `{}` but I think it is serialized as `:null`. Compare:

```emacs-lisp
(jsonrpc--json-encode '(:autopep8 (:enabled t)))            ; "{\"autopep8\":{\"enabled\":true}}"
(jsonrpc--json-encode '(:autopep8 (:enabled :json-false)))  ; "{\"autopep8\":{\"enabled\":false}}"
(jsonrpc--json-encode '(:autopep8 (:enabled nil)))          ; "{\"autopep8\":{\"enabled\":null}}"
(jsonrpc--json-encode '(:autopep8 (:enabled [])))           ; "{\"autopep8\":{\"enabled\":[]}}"

(json-serialize '(:autopep8 (:enabled t)))            ; "{\"autopep8\":{\"enabled\":true}}"
(json-serialize '(:autopep8 (:enabled :json-false)))  ; (wrong-type-argument json-value-p :json-false)
(json-serialize '(:autopep8 (:enabled nil)))          ; "{\"autopep8\":{\"enabled\":{}}}"
(json-serialize '(:autopep8 (:enabled [])))           ; "{\"autopep8\":{\"enabled\":[]}}"
```
According to this [Emacs manual page](https://www.gnu.org/software/emacs/manual/html_node/elisp/Parsing-JSON.html) (which the eglot manual links to also) `nil` indeed represents the empty JSON object. But which is it from the above?

It also states, the JSON `null` is represented by the `:null` symbol and that the JSON array `[]` is represented by Lisp vectors so the `[]` symbol with the caveat that Emacs Lisp vector elements should not be comma separated as they are in a JSON array.

Now `jsonrpc.el` seems to set a `:null-object` keyword symbol to `nil` as well. However, perhaps in keeping with that same Emacs manual page `jsonrpc.el` should set `:null-object` to something like `:json-null` (similar to how it sets `:false-object` to `:json-false`) rather than `nil` to produce `:null` and use `nil` to produce `:{}` as the eglot manual suggests?

That is, right now I think it is:

```
;; Emacs Lisp  |  JSON
;; -------------------------------
;; t           | :true i.e. true
;; :json-false | :false i.e. false
;; nil         | :null i.e. null
;; []          | :[] i.e. [], an empty array
;; ?           | :{} i.e. {}, an empty object
```
but I'm thinking should it be:
```
;; Emacs Lisp  |  JSON
;; -------------------------------
;; t           | :true i.e. true
;; :json-false | :false i.e. false
;; :json-null  | :null i.e. null
;; []          | :[] i.e. [], an empty array
;; nil         | :{} i.e. {}, an empty object
```
(Note, I raised these points in [this PR discussion](https://github.com/joaotavora/eglot/pull/790#discussion_r968444946) previously but I think they got lost in that long thread on a complicated subject. I'm asking again here as I would like to clarify these points in a complete  `.dir-locals.el` file I wish to provide here for the `python-lsp-server`.)

Thanks for your time.

**Comment by joaotavora** [comment, marked as answer]
@doolio thanks for raising this.  We need to first ensure that the manual accurately describes the current behaviour and only then think if the current behaviour can be extended in a backward-compatible manner.

* It seems that the fact that Elisp `nil` maps to JSON `null`.  So that is indeed a bug in the manual that must be corrected.  You can offer to correct it in the MANUAL.md, but it will be a very short lived correction as the manual is undergoing writing to TexInfo.  I will apply the correction there ASAP.

* The Elisp object that guaranteedly maps to JSON `{}` is the empty hash table.  This object is available in the internal Eglot constant  `eglot--{}`.  I will make this constant external so that programs may use it.  In the meantime, the form `(make-hash-table)` evaluates to such an object.

* The Elisp expression that maps to JSON `[]` is the empty Elisp array `[]`.  This is also missing in the manual.  Again, I will add it to the Texinfo version, and you can, if you wish submit the short-lived correction to MANUAL.md

> However, perhaps in keeping with that same Emacs manual page jsonrpc.el should set :null-object to something like :json-null (similar to how it :false-object to :json-false) rather than nil to produce :null and use nil to produce :{} as the eglot manual suggests?

That may indeed have been a better decision back when `jsonrpc.el` was created (by me), but if it were to do this now as you suggest, it would break its users (the most prominent of which being Eglot, but it's not the only one).  This can't be.

**Reply by doolio** [reply]
Always, happy to contribute even if short-lived. I've submitted PR #1089 for consideration. Please let me know if you want it changed in any way.

> The Elisp object that guaranteedly maps to JSON `{}` is the empty hash table. This object is available in the internal Eglot constant `eglot--{}`. I will make this constant external so that programs may use it.

Ah, thanks for the explanation. Yes, that would be good as certainly the `python-lsp-server` has at least one ~~default~~ setting ~~set to `{}`~~ of type `object` namely `pylsp.plugins.jedi.env_vars`.
 
> In the meantime, the form `(make-hash-table)` evaluates to such an object. 

OK, but I get the following:

```
(jsonrpc--json-encode '(:autopep8 (:enabled (make-hash-table))))  ; (void-function jsonrpc--json-encode)
(json-serialize '(:autopep8 (:enabled (make-hash-table))))  ; (wrong-type-argument consp nil)
```

> That may indeed have been a better decision back when `jsonrpc.el` was created (by me), but if it were to do this now as you suggest, it would break its users (the most prominent of which being Eglot, but it's not the only one). This can't be.

I see. I did not appreciate that point.

Edit: corrected one statement.

**Comment by doolio** [comment]
@joaotavora firstly congratulations for Eglot being added to Emacs core.

Considering our discussion above and reading the newly published [manual](https://joaotavora.github.io/eglot/#Customizing-Eglot) I note the manual doesn't list the JSON array value `[]` though it does include the other corrections. There is also a typo in "keywork-value". As for the newly exposed `eglot-{}` value I get the following error:

```
(jsonrpc--json-encode '(:autopep8 (:enabled eglot-{})))     ; wrong-type-argument: json-value-p, eglot-{}
```

(FYI, I've just received the forms for copyright assignment. I will be sending them back shortly. I'm only learning to programme but perhaps I can make some meaningful contributions in future.)

**Reply by joaotavora** [reply]
You have to use 

```
`(:autopep8 (:enabled ,eglot-{})))
```

or

```
(list :autopep8 (list :enabled eglot-{}))
```

Learn about how Lisp quoting `'` and backquoting works.

As to the omissions in the manual, you can submit a patch to emacs.  If the patch is small, you don't need the copytight assignment, I think.

**Reply by doolio** [reply]
~~Can't seem to get either form to work.~~

~~Edit: Adding a more complete form to show where this lisp object value is used in my case.~~

```emacs-lisp
(jsonrpc--json-encode `(:pylsp (:plugins (:autopep8 (:enabled t) :jedi (:env_vars ,eglot-{} :environment nil :extra_paths [])))))
```

```emacs-lisp
(jsonrpc--json-encode (list :pylsp (list :plugins (list :autopep8 (list :enabled t) :jedi (list :env_vars eglot-{} :environment nil :extra_paths [])))))
```
Ignore me. Sorry for the noise. Both forms seem to be working now.

**Reply by doolio** [reply]
Actually, despite the above forms working, when included in the `.dir-locals.el` file the `eglot-{}` lisp object is not evaluated and produces the following error:

```emacs-lisp
jsonrpc--json-encode: Wrong type argument: json-value-p, eglot-{}
```
